### ISSUE TITLE
[func-style] Preferring function expressions, but allowing typescript-specific function overloads

### ISSUE THREAD

**Issue by kachkaev** [typescript-eslint/typescript-eslint#1236]
Our teams use [`func-style`](https://eslint.org/docs/rules/func-style) with default options, which means that we stick with function expressions instead of declarations everywhere. We find this better than seeing a mix of declarations (`function` keyword) and inline arrow functions (no `function` keyword), which would be the case if we followed `typescript-eslint` recommendations. See a footnote on `func-style` in [ROADMAP.md](https://github.com/typescript-eslint/typescript-eslint/blob/ca41dcf6c0fbfc19975b18ffb5b44c0cbe8adb06/packages/eslint-plugin/ROADMAP.md), suggesting to use `"func-style": ["error", "declaration", { "allowArrowFunctions": true }]`.
 
The only problematic case we have is when we do [TypeScript-specific  function overloads](https://www.typescriptlang.org/docs/handbook/functions.html#overloads): 

```TS
// this is the function style we prefer in general
const normalFunc = (x: string) => x + x

// this is what we have to write when there is a need for an overload
function overloadedFunc(x: string): string
function overloadedFunc(x: number): number
function overloadedFunc(x: any): any {
 return x + x
}
```

Unless we change `func-style`  options, we get the following error in the second case:

```
Expected a function expression. eslint(func-style)
```

Function overloading is an advanced TypeScript feature, which we aim to use rarely. Changing hundreds of files because of switching to `["error", "declaration", { "allowArrowFunctions": true }]` is not the price we are ready to pay. In addition to a large `git diff`, the code will become more confusing for JavaScript newcomers – they would need to get exposed into subtle differences between declarations and expressions. Stylistic consistency will suffer as well.

For now, we just use `/* eslint-disable func-style */` in files that contain function overloads, but that's not ideal. Here is what we would prefer instead:

**Expected Result**

```diff
 {
   "rules": {
-    "func-style": "error",
+    "@typescript-eslint/func-style": ["error", { allowOverloads: true }],
   }
 }
```

What do you think of cloning the original rule to support overloads in codebases when function expression are more preferred? This would be a similar case to [require-await](https://eslint.org/docs/rules/require-await) → [@typescript-eslint/require-await](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/require-await.md). What would be the arguments against implementing such a rule?

Perhaps, there is even no need for an option – `allowOverloads` can just always be `true`.

```diff
 {
   "rules": {
-    "func-style": "error",
+    "@typescript-eslint/func-style": "error",
   }
 }
```

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.4.0` |
| `@typescript-eslint/parser`        | `2.4.0` |
| `TypeScript`                       | `3.7.2` |
| `ESLint`                           | `6.5.1` |
| `node`                             | `12.13.0` |
| `npm`                              | `6.12.0` |

**Comment by bradzacher** [comment]
Note that the footnote in the roadmap is mentioning how to migrate from tslint, not the best options to use for typescript.

That's the guidance for that entire document - it is intended to list eslint-alternatives to tslint rules, to help people migrate configs.

---

I don't have an issue with extending base rules to support ts features.
It's generally the best way for us to handle things.

If this is important to you, happy to accept a PR.

**Comment by phaux** [comment]
I'm pretty sure you can do
```ts
type Foo = {
  (x: string): string,
  (x: number): number,
  (x: any): any,
}

const foo: Foo = x => x + x
```

**Comment by lukevella** [comment]
@phaux Unfortunately this doesn't work if your function is typed to anything other then `any`.

```typescript
interface Foo {
  (x: string): string
  (x: number): number
}

const foo: Foo = (x: string | number) => x
```

Results in the following error:

```
Type '(x: string | number) => string | number' is not assignable to type 'Foo'.
  Type 'string | number' is not assignable to type 'string'.
    Type 'number' is not assignable to type 'string'. ts(2322)
```

**Comment by JoshuaKGoldberg** [comment]
Good news! ESLint core added support for TypeScript AST syntax in their rules. See #11014.

Since we're deprecating our syntax-only extension rules (#11014), this issue is no longer something to be tackled in typescript-eslint core. Cheers!

N.b. for `func-style` specifically, https://github.com/eslint/eslint/issues/19719 -> https://github.com/eslint/eslint/pull/19754 added an `allowTypeAnnotation` option to the core [`func-style` rule](https://eslint.org/docs/latest/rules/func-style). It's not quite what this issue is looking for, but it's an example.
